contracts/core/KernelGateway.sol

```diff
@@ -0,0 +1,540 @@
+        );
+
+        require(
+            _updatedValidators.length == _updatedWeights.length,
+            "The lengths of the addresses and weights arrays must be identical."
+        );
+
+        require(
+            originBlockStore.latestBlockHeight() >= _originBlockHeight,
+            "The block containing the state root must be finalized."
+        );
+
+        require(
+            _storageBranchRlp.length > 0,
+            "The storage branch rlp must not be zero."
+        );
+
+        require(
+            _accountRlp.length > 0,
+            "The RLP encoded account must not be zero."
+        );
+
+        require(
+            _accountBranchRlp.length > 0,
+            "The RLP encoded account node path must not be zero."
+        );
+
+        require(
+            _parent != bytes32(0),
+            "Parent hash must not be zero."
+        );
+
+        require(
+            _auxiliaryBlockHash != bytes32(0),
+            "Auxiliary block hash must not be zero."
+        );
+
+        bytes32 kernelHash = MetaBlock.hashKernel(
+            _height,
+            _parent,
+            _updatedValidators,
+            _updatedWeights
+        );
+
+        validateKernel(
+            _height,
+            _parent,
+            kernelHash,
+            _auxiliaryBlockHash
+        );
+
+        require(
+            kernels[kernelHash].height == 0,
+            "Kernel must not exist."
+        );
+
+        bytes32 stateRoot = originBlockStore.stateRoot(_originBlockHeight);
+
+        // State root should be present for the block height
+        require(
+            stateRoot != bytes32(0),
+            "The State root must not be zero."
+        );
+
+        bytes32 storageRoot = getStorageRoot(
+            _accountRlp,
+            _accountBranchRlp,
+            encodedOriginCorePath,
+            stateRoot
+        );
+
+        /* Verify merkle proof for the existence of data */
+        require(
+            verify(
+                keccak256(abi.encodePacked(kernelHash)),
+                storagePath,
+                _storageBranchRlp,
+                storageRoot
+            ),
+            "Storage proof must be verified."
+        );
+
+        // Store the new kernel object.
+        kernels[kernelHash] = MetaBlock.Kernel(
+            _height,
+            _parent,
+            _updatedValidators,
+            _updatedWeights
+        );
+
+        // Store the activation height for the reported kernel.
+        openKernelActivationHeight =
+        auxiliaryBlockStore.getCurrentDynasty().add(2);
+
+        // Store the kernel hash for activation
+        openKernelHash = kernelHash;
+
+        emit OpenKernelProven(
+            originBlockStore.getCoreIdentifier(),
+            auxiliaryBlockStore.getCoreIdentifier(),
+            _height,
+            _parent,
+            kernelHash,
+            auxiliaryBlockStore.getCurrentDynasty().add(2)
+        );
+
+        success_ = true;
+
+    }
```

Comment:
Very long function. Is it possible to split it to improve readability?

Reply:
It is already split, the number of lines appears a lot because almost every statement has 4 params and due to style guidelines, it's increasing the number of lines as its placed one after the another.

I can only move the validation code now. What you suggest ?

Reply:
It is "19" lines of code actually with mostly requires. I think it is fine.